fdbclient/ManagementAPI.actor.cpp

```diff
@@ -1587,6 +1589,69 @@ ACTOR Future<Void> excludeServers(Database cx, vector<AddressExclusion> servers,
 				wait(tr.commit());
 				return Void();
 			} catch (Error& e) {
+				TraceEvent("ExcludeServersError").error(e, true);
+				wait(tr.onError(e));
+			}
+		}
+	}
+}
+
+void excludeLocalities(Transaction& tr, std::unordered_set<std::string>& localities, bool failed) {
+	tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+	tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+	tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+	tr.setOption(FDBTransactionOptions::USE_PROVISIONAL_PROXIES);
+	std::string excludeVersionKey = deterministicRandom()->randomUniqueID().toString();
+	auto localityVersionKey = failed ? failedLocalityVersionKey : excludedLocalityVersionKey;
+	tr.addReadConflictRange(singleKeyRange(localityVersionKey)); // To conflict with parallel includeLocalities
+	tr.set(localityVersionKey, excludeVersionKey);
+	for (auto& l : localities) {
+		if (failed) {
+			tr.set(encodeFailedLocalityKey(l), StringRef());
+		} else {
+			tr.set(encodeExcludedLocalityKey(l), StringRef());
+		}
+	}
+	TraceEvent("ExcludeLocalitiesCommit")
+		.detail("Localities", describe(localities))
+		.detail("ExcludeFailed", failed);
+}
+
+ACTOR Future<Void> excludeLocalities(Database cx, std::unordered_set<std::string> localities, bool failed) {
+	if (cx->apiVersionAtLeast(700)) {
+		state ReadYourWritesTransaction ryw(cx);
+		loop {
```

Comment:
This should probably have an exponential backoff / jitter for the retries. I sincerely doubt that management API actions could ever overwhelm the database, but it is good practice to always add a jittered exponential backoff to any retried action against a distributed system. Some relevant reading:

https://sre.google/sre-book/addressing-cascading-failures/
https://sre.google/sre-book/handling-overload/

Reply:
In this loop, `ryw.onError(e)` in the catch block provides the backoff mechanism.

Reply:
Do you happen to have a code pointer? When I read through the implementation I didn't see any backoff logic.

Reply:
https://github.com/apple/foundationdb/blob/5f04c2c7f5aad38581c43a28d7eb2ea23881d6f3/fdbclient/NativeAPI.actor.cpp#L5227